pkg/controller/source/source.go

```diff
@@ -46,18 +47,115 @@ type Source interface {
 	Start(eventhandler.EventHandler, workqueue.RateLimitingInterface, ...predicate.Predicate) error
 }
 
+var _ Source = &ChannelSource{}
+
 // ChannelSource is used to provide a source of events originating outside the cluster
 // (eh.g. GitHub Webhook callback).  ChannelSource requires the user to wire the external
 // source (eh.g. http handler) to write GenericEvents to the underlying channel.
-type ChannelSource chan event.GenericEvent
+type ChannelSource struct {
+	// once ensures the event distribution goroutine will be performed only once
+	once sync.Once
+	// source is the source channel to fetch GenericEvents
+	source chan event.GenericEvent
+	// stop is to end ongoing goroutine, and close the channels
+	stop <-chan struct{}
+	// dest is the destination channels of the added event handlers
+	dest []chan event.GenericEvent
+	// destLock is to ensure the destination channels are safely added/removed
+	destLock sync.Mutex
+}
+
+var _ inject.SourceChannel = &ChannelSource{}
+
+// InjectSourceChannel is internal should be called only by the Controller.
+// It is used to inject the source event channel initialized by the ControllerManager.
+func (cs *ChannelSource) InjectSourceChannel(c chan event.GenericEvent) error {
+	if cs.source == nil {
+		cs.source = c
+	}
+
+	return nil
+}
+
+var _ inject.Stop = &ChannelSource{}
 
-var _ Source = ChannelSource(make(chan event.GenericEvent))
+// InjectStop is internal should be called only by the Controller.
+// It is used to inject the stop channel initialized by the ControllerManager.
+func (cs *ChannelSource) InjectStop(stop <-chan struct{}) error {
+	if cs.stop == nil {
+		cs.stop = stop
+	}
+
+	return nil
+}
 
 // Start implements Source and should only be called by the Controller.
-func (ks ChannelSource) Start(
+func (cs *ChannelSource) Start(
 	handler eventhandler.EventHandler,
 	queue workqueue.RateLimitingInterface,
 	prct ...predicate.Predicate) error {
+	// source should have been injected before Start was called.
+	if cs.source == nil {
+		return fmt.Errorf("must call DoSourceChannel on ChannelSource before calling Start")
+	}
+
+	// stop should have been injected before Start was called
+	if cs.stop == nil {
+		return fmt.Errorf("must call DoStop on ChannelSource before calling Start")
+	}
+
+	cs.once.Do(func() {
+		// Distribute GenericEvents to all EventHandler / Queue pairs Watching this source
+		go func() {
+			for {
+				select {
+				case <-cs.stop:
+					// Close source channel.
+					// TODO: should we close the source channel here?
```

Comment:
Not sure if we need to close the source channel here.
Maybe users meant to use it somewhere out of the controller, and handle its lifecycle themselves?

Reply:
I think the lifecycle of the source channel should be completely managed by the user.  We shouldn't create it or close it for them.